[rrfs-mpas-jedi] Add infrastructure for nonvar cloud analysis

ShawnMurdzek-NOAA opened the pull request:
## DESCRIPTION OF CHANGES: 
This PR adds the necessary infrastructure to run the nonvariational cloud analysis within the MPAS-JEDI version of rrfs-workflow. In particular, three new tasks are added:
- `proc_bufr_nonvar`: Process NASA LaRC, lightning, and METAR BUFR files
- `refmosaic_nonvar`: Process NSSL MRMS reflectivity
- `cloudanalysis_nonvar`: Perform the nonvariational cloud analysis

Similar to RRFSv1, the nonvariational cloud analysis sits between the main data assimilation task (in this case, `jedivar`) and the model forecast task. Many of the namelist settings for the nonvariational cloud analysis tasks are designed to mimic RRFSv1. Future testing using retrospective experiments will refine these options.

The nonvariational cloud analysis can be turned on by setting `DO_NONVAR_CLOUD_ANA=true`. By default, this option is turned on for `exp.conus3km`, but is turned off for all other workflow configurations. This reflects the fact that the nonvariational cloud analysis is currently not compatible with the 12-km CONUS deterministic configuration or any ensemble configuration. A future PR will extend the nonvariational cloud analysis to the 12-km CONUS deterministic configuration (this will require some small changes to [RRFS_UTILS](https://github.com/NOAA-GSL/RRFS_UTILS) that slightly increase the size of the map projection used by the nonvariational cloud analysis). Although it has not been explicitly tested, the nonvariational cloud analysis is expected to work with the 3.5-km south deterministic configuration.

*Technical note*: The nonvariational cloud analysis is only run during warm starts. This is because during a cold start, the `cldfrac` field, which is required for `cloudanalysis_nonvar`, is missing. The `cloudanalysis_nonvar` task is still defined for cold starts, but the task does not do anything. Thus, when running `exp.conus3km`, the first cycle that will actually run the full nonvariational cloud analysis is the 01 UTC cycle.

## TESTS CONDUCTED: 
- Updated code compiles on both Gaea C6 and Ursa
- Successfully generated a workflow using the `exp.conus3km` configuration with `DO_NONVAR_CLOUD_ANA=true`. The resulting workflow runs without error for multiple cycles (3+) on both Ursa and Gaea C6.
- A comparison between `exp.conus3km` RRFS runs with `DO_NONVAR_CLOUD_ANA=true` and `DO_NONVAR_CLOUD_ANA=false` on Gaea C6 shows differences in the `QCLOUD` field, which indicates that the nonvariational cloud analysis is altering the cloud fields (as expected).
- Can successfully generate a workflow with the `exp.conus12km` configuration with `DO_NONVAR_CLOUD_ANA=true` on Gaea C6, but `proc_bufr_nonvar` crashes when this workflow is run owing to the map projection not being large enough in `metarcld.fd`. This will be fixed in future PRs.

ShawnMurdzek-NOAA commented on workflow/sideload/launch.sh:
```diff
@@ -72,6 +72,15 @@ case ${task_id} in
+    module purge
+    module use "${HOMErrfs}/sorc/RRFS_UTILS/modulefiles"
+    if [[ ${MACHINE} == "gaea" ]]; then
+      module load "build_gaeaC6_intel"
```
Great, thanks! That will certainly simplify things

ShawnMurdzek-NOAA commented:
> For naming the tasks (or jobs), I would recommend we use simplified forms if possible (i.e. without ambiguity). In particular, NCO strongly discourage using verbs in task names if possible (for example, we don't use `process_ioda_bufr`, but use `ioda_bufr` directly).
> 
> If we treat all these tasks as belonging to the `cldana` group, we may use the following task names:
> 
> ```
> cldana_nonvar
> cldana_bufrobs
> cldana_reflobs
> ```
> 
> If we treat all these tasks as belonging to the `nonvar` group, we may use the following task names:
> 
> ```
> nonvar_cldana
> nonvar_bufrobs
> nonvar_reflobs
> ```
> 
> Thanks!

@guoqing-noaa thanks for the suggestion! I just changed the task names to `nonvar_cldana`, `nonvar_bufrobs`, and `nonvar_reflobs`. I created a new workflow using the updated code on Gaea C6 and ran it for a few cycles, and everything appears to still be working as expected.

guoqing-noaa commented on workflow/config_resources/config.base:
```diff
@@ -46,6 +46,13 @@ export NODES_IODA_BUFR=${NODES_IODA_BUFR:-"<nodes>1:ppn=1</nodes>"}
 # ioda_mrms_refl
 export NODES_IODA_MRMS_REFL=${NODES_IODA_MRMS_REFL:-"<nodes>1:ppn=20</nodes>"}
 
+# proc_bufr_nonvar
```
Line 49 comment can be updated to "nonvar_bufrobs"

ShawnMurdzek-NOAA replied:
Updated comment, but did not add `NODES_NONVAR_BUFROBS`

guoqing-noaa commented on scripts/exrrfs_nonvar_reflobs.sh:
```diff
@@ -0,0 +1,140 @@
+    break
+  fi
+  s=0
+  while [[ $s -le 59 ]]; do
```
Suggest using `((...))` for numerical comparision:
`while (( s <= 59); do`

guoqing-noaa commented on scripts/exrrfs_nonvar_reflobs.sh:
```diff
@@ -0,0 +1,140 @@
+      nsslfile1="*${mrms}_*_${YYYY}${MM}${DD}-${HH}${min}*.${obs_appendix}"
+      numgrib2=$(find ${NSSL}/${nsslfile1} -maxdepth 1 -type f | wc -l)
+      echo "Number of GRIB-2 files: ${numgrib2}"
+      if [ "${numgrib2}" -ge 10 ] && [ ! -e filelist_mrms ]; then
```
suggest changing to:
`if (( numgrib2 >= 10)) && [[ ! -s filelist_mrms ]]; then`

guoqing-noaa commented on scripts/exrrfs_nonvar_reflobs.sh:
```diff
@@ -0,0 +1,140 @@
+fi
+
+if [ -s filelist_mrms ]; then
+  if [ "${obs_appendix}" == "grib2.gz" ]; then
```
Suggest using `[[...]]`

guoqing-noaa commented on workflow/config_resources/config.base:
```diff
@@ -46,6 +46,13 @@ export NODES_IODA_BUFR=${NODES_IODA_BUFR:-"<nodes>1:ppn=1</nodes>"}
+# proc_bufr_nonvar
+export WALLTIME_NONVAR_BUFROBS=${WALLTIME_NONVAR_BUFROBS:-"00:25:00"}
+
+# refmosaic_nonvar
```
Suggest changing to `nonvar_reflobs`

guoqing-noaa commented on workflow/rocoto_funcs/nonvar_cldana.py:
```diff
@@ -0,0 +1,60 @@
+    dependencies = f'''
+  <dependency>
+  <and>{timedep}{prep_ic_dep}
+    {jedidep}
```
Sorry that `    {jedidep}` will leave an empty line if we don't DO_JEDI. I have corrected this in `fcst.py` in a recent PR. Could you correct it here as well?
i.e. adding `\n    ` to `jedidep` in lines 40, 42 and 44 and then move `{jedidep}` to the end of line 52.
Thanks!

ShawnMurdzek-NOAA commented on fix/cloudanalysis/prepobs_prep_RAP.bufrtable:
Thanks for pointing this out! My vote would be to not generate the output BUFR files because they are not needed. Unfortunately, there is currently no way to easily turn off the creation of BUFR files in the nonvariational cloud analysis programs. I can add namelist options to RRFS_UTILS in the future so that we can easily turn off the creation of output BUFR files. At that point, we can remove this file.

ShawnMurdzek-NOAA commented on workflow/setup_rocoto.py:
```diff
@@ -28,6 +28,14 @@
+    if mesh not in ["conus3km", "south3.5km", "conus12km"]:
+        print(f'{mesh} is not compatible with the nonvar cloud analysis')
+        print('Please set DO_NONVAR_CLOUD_ANA=false and try again')
+        exit()
```
Added

guoqing-noaa commented:
> @guoqing-noaa I think I addressed all of your comments. Please let me know if I missed something! I confirmed that the updated code runs on Gaea C6 for at least 2 cycles without any of the nonvariational cloud analysis tasks crashing.

@ShawnMurdzek-NOAA Thanks a lot for addressing all the comments!

Sorry for my picky comment on `fix/cloudanalysis/prepobs_prep_RAP.bufrtable`, if we will NOT use those bufr output files in any ways at the moment, let's avoid bringing this file into rrfs-workflow.

I created a PR at https://github.com/NOAA-GSL/RRFS_UTILS/pull/3. Will this PR help address this issue?
Thanks!

ShawnMurdzek-NOAA commented:
> > @guoqing-noaa I think I addressed all of your comments. Please let me know if I missed something! I confirmed that the updated code runs on Gaea C6 for at least 2 cycles without any of the nonvariational cloud analysis tasks crashing.
> 
> @ShawnMurdzek-NOAA Thanks a lot for addressing all the comments!
> 
> Sorry for my picky comment on `fix/cloudanalysis/prepobs_prep_RAP.bufrtable`, if we will NOT use those bufr output files in any ways at the moment, let's avoid bringing this file into rrfs-workflow.
> 
> I created a PR at [NOAA-GSL/RRFS_UTILS#3](https://github.com/NOAA-GSL/RRFS_UTILS/pull/3). Will this PR help address this issue? Thanks!

@guoqing-noaa No worries! I just updated the RRFS_UTILS hash and removed the fix files. That should address your concerns
